src/enclave/Enclave/BroadcastNestedLoopJoin.cpp

```diff
@@ -0,0 +1,52 @@
+#include "BroadcastNestedLoopJoin.h"
+
+#include "ExpressionEvaluation.h"
+#include "FlatbuffersReaders.h"
+#include "FlatbuffersWriters.h"
+#include "common.h"
+
+#include <iostream>
+
+/** C++ implementation of a broadcast nested loop join.
+ * Which row is broadcast and which row is not is handled in Scala.
+ * DOES NOT rely on rows to be tagged primary or secondary, and that
+ * assumption will break the implementation. In other words,
+ * outer is always left and inner is always right.
+ */
+void broadcast_nested_loop_join(
+  uint8_t *join_expr, size_t join_expr_length,
+  uint8_t *outer_rows, size_t outer_rows_length,
+  uint8_t *inner_rows, size_t inner_rows_length,
+  uint8_t **output_rows, size_t *output_rows_length) {
+
+  FlatbuffersJoinExprEvaluator join_expr_eval(join_expr, join_expr_length);
+  const tuix::JoinType join_type = join_expr_eval.get_join_type();
+
+  RowReader outer_r(BufferRefView<tuix::EncryptedBlocks>(outer_rows, outer_rows_length));
+  RowWriter w;
+
+  while (outer_r.has_next()) {
+    const tuix::Row *outer = outer_r.next();
+    bool o_i_match = false;
+
+    RowReader inner_r(BufferRefView<tuix::EncryptedBlocks>(inner_rows, inner_rows_length));
+    const tuix::Row *inner;
+    while (inner_r.has_next()) {
+      inner = inner_r.next();
+      o_i_match |= join_expr_eval.eval_condition(outer, inner);
+    }
+
+    switch(join_type) {
+      case tuix::JoinType_LeftAnti:
```

Comment:
Can you support general left existence join here, with LeftSemi as well?

Reply:
LeftSemi done, but it looks like left existence is not supported? https://github.com/apache/spark/blob/429f8af9b683935151c2379bc80b27162cd1c8bf/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala#L90